design/acme-orders-challenges-crd.md

```diff
@@ -0,0 +1,388 @@
+	// TLS certificate for this order.
+	// It will be populated automatically once the order has completed successfully
+	// and the certificate is available for retrieval.
+	// +optional
+	CertificateURL string `json:"certificateURL,omitempty"`
+
+	// State contains the current state of this Order resource.
+	// States 'success' and 'expired' are 'final'
+	State State `json:"state"`
+
+	// Reason optionally provides more information about a why the order is in
+	// the current state.
+	Reason string `json:"reason"`
+
+	// Challenges is a list of ChallengeSpecs for Challenges that must exist
+	// and be in a 'valid' state in order to complete this order.
+	Challenges []ChallengeSpec `json:"challenges,omitempty"`
+
+	// FailureTime stores the time that this order failed.
+	// This is used to influence garbage collection and back-off.
+	// The order resource will be automatically deleted after 30 minutes has
+	// passed since the failure time.
+	// +optional
+	FailureTime *metav1.Time `json:"failureTime,omitempty"`
+}
+
+// State represents the state of an ACME resource, such as an Order.
+// The possible options here map to the corresponding values in the
+// ACME specification.
+// Full details of these values can be found there.
+// Clients utilising this type **must** also gracefully handle unknown
+// values, as the contents of this enumeration may be added to over time.
+type State string
+
+const (
+	// Unknown is not a real state as part of the ACME spec.
+	// It is used to represent an unrecognised value.
+	Unknown State = ""
+
+	// Valid signifies that an ACME resource is in a valid state.
+	// If an Order is marked 'valid', all validations on that Order
+    // have been completed successfully and the order has been finalized
+    // with a CSR.
+    // You should perform a GET on the 'CertificateURL' in order to retrieve
+    // the existing certificate.
+	// This is a final state.
+	Valid State = "valid"
+
+	// Ready signifies that an ACME resource is in a ready state.
+    // If an Order is marked 'Ready', the order is ready to be finalized.
+    // The order must be finalized with a CSR, and once finalization succeeds,
+    // the order will transition into a 'valid' state.
+	// This is a transient state.
+	Ready State = "ready"
+
+	// Pending signifies that an ACME resource is still pending and is not yet ready.
+	// If an Order is marked 'Pending', the validations for that Order are still in progress.
+	// This is a transient state.
+	Pending State = "pending"
+
+	// Processing signifies that an ACME resource is being processed by the server.
+	// If an Order is marked 'Processing', the validations for that Order are currently being processed.
+	// This is a transient state.
+	Processing State = "processing"
+
+	// Failed signifies that an ACME resource has failed for some reason.
+	// If an Order is marked 'Failed', one of its validations may have failed for some reason.
+	// This is a final state.
+	Failed State = "failed"
```

Comment:
The comment on the `State` type indicate the states correspond to values in the ACME specification but there is no "failed" state in the ACME draft. Should this be using "invalid" instead?